tests/test_pyxis.py

```diff
@@ -0,0 +1,426 @@
+                          'data.bundle_path']
+        expected_params = {'param1': 1, 'include': ','.join(include_fields)}
+        expected_out = {"path/to/registry:v4.5": self.bundles[0:2],
+                        "path/to/registry:v4.6": self.bundles[2:]}
+
+        self.assertEqual(out, expected_out)
+        call_count = 0
+        for index in self.indices:
+            if not index.get('path'):
+                continue
+            call_count += 1
+            expected_params.update({
+                "filter": f"source_index_container_path=={index['path']}"})
+            page_copy.assert_any_call("operators/bundles", expected_params)
+        self.assertEqual(page.call_count, call_count)
+
+    @patch('freshmaker.pyxis.Pyxis._get_bundles_per_index_image')
+    def test_get_latest_bundles(self, get_bundles_per_indices):
+        get_bundles_per_indices.return_value = {
+            "path/to/registry:v4.5": self.bundles[0:2],
+            "path/to/registry:v4.6": self.bundles[2:]
+        }
+
+        out = self.px.get_latest_bundles(self.indices)
+        # we expect 0 and 3 bundles to be filtered out, because of older versions
+        expected_out = self.bundles[1:3] + self.bundles[4:]
+        self.assertEqual(out, expected_out)
+        get_bundles_per_indices.assert_called_once()
+
+    @patch('freshmaker.pyxis.Pyxis._get_bundles_per_index_image')
+    def test_get_latest_bundles_invalid_version(self, get_bundles_per_indices):
+        # set invalid version
+        for bundle in self.bundles:
+            bundle['version'] = "InvalidVersion"
+        get_bundles_per_indices.return_value = {
+            "path/to/registry:v4.5": self.bundles[0:2],
+            "path/to/registry:v4.6": self.bundles[2:]
+        }
+
+        with self.assertLogs("freshmaker", level="WARNING"):
+            bundles = self.px.get_latest_bundles(self.indices)
+            self.assertEqual(bundles, [])
+            get_bundles_per_indices.assert_called_once()
+
+    @patch('freshmaker.pyxis.Pyxis._pagination')
+    def test_get_digests_by_nvrs(self, page):
+        page.return_value = self.images
+        digests = self.px.get_digests_by_nvrs(['s2i-1-2'])
+
+        my_digests = ['sha256:1112', 'sha256:4444', 'sha256:3333', 'sha256:2222']
+        self.assertEqual(digests, set(my_digests))
+        page.assert_called_once_with(
+            'images/nvr/s2i-1-2',
+            {'include': 'data.brew,data.repositories'}
+        )
+
+    def test_filter_bundles_by_related_image_digests(self):
+        digests = ['sha256:111', 'sha256:bbb', 'sha256:ddd']
+        new_bundles = self.px.filter_bundles_by_related_image_digests(
+            digests, self.bundles)
+
+        expected_bundles = [self.bundles[0], self.bundles[2], self.bundles[4]]
+        self.assertListEqual(new_bundles, expected_bundles)
+
+    @patch('freshmaker.pyxis.Pyxis._pagination')
+    def test_get_images_by_digests(self, page):
+        my_pagination = self.copy_call_args(page)
```

Comment:
Why is this needed?

Reply:
Because of this https://docs.python.org/3/library/unittest.mock-examples.html#coping-with-mutable-arguments problem